### PR TITLE
Introduce PayloadConverter

### PR DESCRIPTION
**Pull request by alexshtin**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -115,4 +115,4 @@ fmt:
 clean:
 	rm -rf $(BUILD)
 
-check: lint vet errcheck staticcheck copyright bins
```
Comment (alexshtin) [inline Makefile:118]: Will enable it later. It gives me 34 errors now.

File: internal/encoded.go [hunk 2]
```diff
@@ -25,19 +25,26 @@
...
 const (
-	metadataEncoding     = "encoding"
-	metadataEncodingRaw  = "raw"
-	metadataEncodingJSON = "json"
+	metadataEncoding          = "encoding"
+	metadataEncodingRaw       = "binary/raw"
+	metadataEncodingJSON      = "json/plain"
+	metadataEncodingNil       = "binary/null"
+	metadataEncodingProtoJSON = "json/protobuf"
+	metadataEncodingProto     = "binary/protobuf"
```
Comment (alexshtin) [inline internal/encoded.go:47]: Idea behind these names that `json/*` can be directly displayed in UI and basically human readable. `binary/` is vice versa.

File: internal/encoded.go [hunk 3]
```diff
@@ -69,45 +76,79 @@ type (
...
+	ErrValueDoesntImplementProtoUnmarshaler = errors.New("value doesn't implement proto.Unmarshaler")
 )
 
 // getDefaultDataConverter return default data converter used by Temporal worker.
+// TODO: remove this func
 func getDefaultDataConverter() DataConverter {
 	return DefaultDataConverter
 }
 
+// NewDataConverter creates new instance of DataConverter.
```
Comment (mfateev) [inline internal/encoded.go:137]: Add comment to explain how and why this list should be ordered.

File: internal/encoded.go [hunk 4]
```diff
@@ -237,36 +269,270 @@ func getEncoding(payload *commonpb.Payload) (string, error) {
...
+		newProtoValue := reflect.New(protoType)                  // is of type i.e. *commonpb.WorkflowType
+		protoMessage = newProtoValue.Interface().(proto.Message) // type assertion will always succeed
+	}
+
+	err := c.unmarshaler.Unmarshal(bytes.NewReader(payload.GetData()), protoMessage)
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrUnableToDecode, err)
+	}
+
+	value.Set(reflect.ValueOf(protoMessage))
```
Comment (alexshtin) [inline internal/encoded.go:468]: This is a voodoo code.

File: internal/encoded.go [hunk 5]
```diff
@@ -237,36 +269,270 @@ func getEncoding(payload *commonpb.Payload) (string, error) {
 	return "", ErrEncodingIsNotSet
 }
 
-func toString(payload *commonpb.Payload) (string, error) {
-	result := ""
+type PayloadConverter interface {
+	// ToPayload single value to payload.
```
Comment (mfateev) [inline internal/encoded.go:273]: Document that it should return nil if the value type is not something it knows how to deal with.

File: internal/encoded.go [hunk 6]
```diff
@@ -237,36 +269,270 @@ func getEncoding(payload *commonpb.Payload) (string, error) {
...
+	ToPayload(value interface{}) (*commonpb.Payload, error)
+	// FromPayload single value from payload.
+	FromPayload(payload *commonpb.Payload, valuePtr interface{}) error
```
Comment (mfateev) [inline internal/encoded.go:276]: Document that it should return an error if valuePtr is of a type it doesn't know how to deal with.

File: encoded/encoded.go [hunk 7]
```diff
@@ -37,18 +37,33 @@ type (
...
 	// DataConverter is used by the framework to serialize/deserialize input and output of activity/workflow
 	// that need to be sent over the wire.
-	// To encode/decode workflow arguments, one should set DataConverter in two places:
-	//   1. Workflow worker, through worker.Options
-	//   2. Client, through client.Options
-	// To encode/decode Activity/ChildWorkflow arguments, one should set DataConverter in two places:
-	//   1. Inside workflow code, use workflow.WithDataConverter to create new Context,
+	// To encode/decode workflow arguments, one should set DataConverter in client, through client.Options.
+	// To encode/decode Activity/ChildWorkflow arguments, one should set DataConverter
+	// inside workflow code, use workflow.WithDataConverter to create new Context,
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply encoded/encoded.go:42]: It is `client.Options` now. Improved comment.

**Pull request merged**